### PR TITLE
Tickers batch processing

### PR DESCRIPTION
**Pull request by vstavetski**
New module which calculate tickers from historical data from cassandra

### CODE REVIEW

File: tickers-batch/src/main/kotlin/fund/cyber/markets/ticker/service/TickerService.kt [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+        }
+
+        tickers = tickers.mergeWith(
+            tickerRepository.find(symbol, epochDay, Date(timestampFromVar), Date(timestampTo), interval)
+        )
+
+        return tickers
+    }
+
+    private fun cacheKey(symbol: String, timestampFrom: Long, timestampTo: Long, interval: Long): String {
```
Comment (arturalbov) [inline tickers-batch/src/main/kotlin/fund/cyber/markets/ticker/service/TickerService.kt]: use data class

File: tickers-batch/src/main/kotlin/fund/cyber/markets/ticker/service/TickerService.kt [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+                        val cachedValue = cache.get(cacheKey(symbol, timestampFromVar, timestampToVar, interval)) ?: mutableListOf()
+                        when {
+                            cachedValue.isNotEmpty() -> Flux.fromIterable(cachedValue)
+                            else -> Flux.empty()
+                        }
+                    }.switchIfEmpty(
+                        tickerRepository
+                            .find(symbol, epochDay, Date(timestampFromVar), Date(timestampToVar), interval)
+                            .map { ticker ->
+                                val key = cacheKey(symbol, timestampFrom, timestampTo, interval)
```
Comment (arturalbov) [inline tickers-batch/src/main/kotlin/fund/cyber/markets/ticker/service/TickerService.kt]: use flatMap and wrap body into Mono.fromCallable{...}

File: tickers-batch/src/main/kotlin/fund/cyber/markets/ticker/service/TokenPriceService.kt [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+@Service
+class TokenPriceService(
+    private val tokenPriceRepository: TokenPriceRepository
+) {
+    private val log = LoggerFactory.getLogger(TokenPriceService::class.java)!!
+
+    fun save(prices : MutableCollection<TokenPrice>) {
+        log.info("Saving token prices. Count: ${prices.size}")
+
+        tokenPriceRepository.saveAll(prices.map { CqlTokenPrice(it) }).collectList().block()
```
Comment (arturalbov) [inline tickers-batch/src/main/kotlin/fund/cyber/markets/ticker/service/TokenPriceService.kt]: better to return Publisher and block in the place where you're really need it

File: tickers-batch/src/main/kotlin/fund/cyber/markets/ticker/service/TradeService.kt [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+        for (epochMinute in epochMinuteFrom..epochMinuteTo) {
+            cqlTrades = cqlTrades.mergeWith(tradeRepository.findByEpochMinute(epochMinute))
+        }
+
+        return cqlTrades
+            .collectList()
+            .defaultIfEmpty(mutableListOf())
+            .block()!!
+            .map { cqlTrade -> cqlTrade.toTrade() }
+            .sortedBy { trade -> trade.timestamp }
```
Comment (arturalbov) [inline tickers-batch/src/main/kotlin/fund/cyber/markets/ticker/service/TradeService.kt]: Make method return Flux and move sort in places where it needed

File: tickers-batch/src/main/kotlin/fund/cyber/markets/ticker/service/TradeService.kt [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+    }
+
+    fun getLastProcessedTimestamp(): Long {
```
Comment (arturalbov) [inline tickers-batch/src/main/kotlin/fund/cyber/markets/ticker/service/TradeService.kt]: Make method return Mono

File: tickers-batch/src/main/kotlin/fund/cyber/markets/ticker/service/TradeService.kt [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+        val lastTimestamp = lastTimestampRepository.findTradeLastTimestamp()
+            .defaultIfEmpty(CqlTradeLastTimestamp(value = START_TIMESTAMP_DEFAULT))
+            .block()
+
+        log.info("Last processed timestamp: ${Date(lastTimestamp!!.value)}")
+
+        return lastTimestamp.value
+    }
+
+    fun updateLastProcessedTimestamp(timestamp: Long) {
```
Comment (arturalbov) [inline tickers-batch/src/main/kotlin/fund/cyber/markets/ticker/service/TradeService.kt]: Make method return Mono

File: storer/src/main/kotlin/fund/cyber/markets/storer/TradeStorer.kt [hunk 7]
```diff
@@ -1,19 +1,23 @@
...
 class TradeStorer(
-    val tradeRepository: TradeRepository
+    val tradeRepository: TradeRepository,
+    val tradeTemporaryRepository: TradeTemporaryRepository
 ) {
 
     @KafkaListener(id = "tradeListener", topicPattern = "TRADES-.*")
     fun tradeListener(trades: List<Trade>) {
         tradeRepository.saveAll(trades.map { trade -> CqlTrade(trade) }).collectList().block()
+        tradeTemporaryRepository.saveAll(trades.map { trade -> CqlTradeTemporary(trade) }).collectList().block()
```
[the remark this answers is not in the document]
Comment (vstavetski) [reply storer/src/main/kotlin/fund/cyber/markets/storer/TradeStorer.kt]: done

File: tickers-batch/src/main/kotlin/fund/cyber/markets/ticker/processor/HistoricalTickerProcessor.kt [hunk 8]
```diff
@@ -0,0 +1,196 @@
...
+        tickerQuote.baseVolume[baseTokenSymbol]!![exchange] = volumeQuoteBCT
+
+        val volumeQuoteBCTall = tickerQuote.baseVolume
+            .getOrPut(baseTokenSymbol, { mutableMapOf() })
+            .getOrPut(Exchanges.ALL, { BigDecimal.ZERO })
+            .plus(priceForQuote.multiply(trade.quoteAmount))
+        tickerQuote.baseVolume[baseTokenSymbol]!![Exchanges.ALL] = volumeQuoteBCTall
+    }
+
+    private fun updateMapOfPrices(trades: List<Trade>) {
```
[the remark this answers is not in the document]
Comment (vstavetski) [reply tickers-batch/src/main/kotlin/fund/cyber/markets/ticker/processor/HistoricalTickerProcessor.kt]: done

File: tickers-batch/src/main/kotlin/fund/cyber/markets/ticker/processor/price/VwapPriceProcessor.kt [hunk 9]
```diff
@@ -0,0 +1,116 @@
...
+
+            // get tickers from epoch 00:00
+            val symbol = ticker.symbol
+            val timestampFrom = closestSmallerMultiply(ticker.timestampFrom, Intervals.DAY)
+            val timestampTo = ticker.timestampTo - ticker.interval
+            val interval = ticker.interval
+
+            log.debug("Getting previous tickers for ${ticker.symbol}")
+
+            val previousTickers = tickerService.findTickersByInterval(symbol, timestampFrom, timestampTo, interval)
```
[the remark this answers is not in the document]
Comment (vstavetski) [reply tickers-batch/src/main/kotlin/fund/cyber/markets/ticker/processor/price/VwapPriceProcessor.kt]: done

File: tickers-batch/src/main/kotlin/fund/cyber/markets/ticker/service/TickerService.kt (hunk 1 above)
Comment (vstavetski) [reply tickers-batch/src/main/kotlin/fund/cyber/markets/ticker/service/TickerService.kt]: done

File: tickers-batch/src/main/kotlin/fund/cyber/markets/ticker/service/TickerService.kt (hunk 2 above)
Comment (vstavetski) [reply tickers-batch/src/main/kotlin/fund/cyber/markets/ticker/service/TickerService.kt]: done

File: tickers-batch/src/main/kotlin/fund/cyber/markets/ticker/service/TokenPriceService.kt (hunk 3 above)
Comment (vstavetski) [reply tickers-batch/src/main/kotlin/fund/cyber/markets/ticker/service/TokenPriceService.kt]: done

File: tickers-batch/src/main/kotlin/fund/cyber/markets/ticker/service/TradeService.kt (hunk 4 above)
Comment (vstavetski) [reply tickers-batch/src/main/kotlin/fund/cyber/markets/ticker/service/TradeService.kt]: done

File: tickers-batch/src/main/kotlin/fund/cyber/markets/ticker/service/TradeService.kt (hunk 5 above)
Comment (vstavetski) [reply tickers-batch/src/main/kotlin/fund/cyber/markets/ticker/service/TradeService.kt]: done

File: tickers-batch/src/main/kotlin/fund/cyber/markets/ticker/service/TradeService.kt (hunk 6 above)
Comment (vstavetski) [reply tickers-batch/src/main/kotlin/fund/cyber/markets/ticker/service/TradeService.kt]: done

**Pull request merged**